dbms/src/Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.cpp

```diff
@@ -0,0 +1,265 @@
+
+            // When highest priority of resource group is less than zero, means RU of all resource groups are exhausted.
+            // Should not take any task from nested task queue for this situation.
+            if (LocalAdmissionController::isRUExhausted(priority))
+                break;
+
+            if (task_queue->empty() || !task_queue->take(task))
+            {
+                LOG_TRACE(
+                    logger,
+                    "take task from nested task_queue of resource group {} failed. task_queue.empty(): {}",
+                    name,
+                    task_queue->empty());
+                // Got here only when task_queue is empty or finished, we try next resource group.
+                // If new task of this resource gorup is submited, the resource_group info will be added again.
+                resource_group_infos.pop();
+                size_t erase_num = resource_group_task_queues.erase(name);
+                RUNTIME_CHECK_MSG(
+                    erase_num == 1,
+                    "cannot erase corresponding TaskQueue for task of resource group {}",
+                    name);
+            }
+            else
+            {
+                LOG_TRACE(
+                    logger,
+                    "schedule task of resource group {} succeed, cur cpu time of MPPTask: {}",
+                    name,
+                    task->getQueryExecContext().getQueryProfileInfo().getCPUExecuteTimeNs());
+                assert(task != nullptr);
+                break;
+            }
+        }
+
+        if (task != nullptr)
+            return true;
+
+        // Check is_finished again for situation when resource_group_infos never insert any resource group.
+        if unlikely (is_finished)
+            return false;
+
+        // Other TaskQueue like MultiLevelFeedbackQueue and IOPriorityQueue will wake up when new task submit or is_finished become true.
+        // But for ResourceControlQueue, when all resource groups's RU are exhausted, will go to sleep, and should wakeup when RU is updated.
+        // But LAC has no way to notify ResourceControlQueue for now, so ResourceControlQueue should wakeup to check if RU is updated or not.
+        cv.wait_for(lock, DEFAULT_WAIT_INTERVAL_WHEN_RUN_OUT_OF_RU);
+        updateResourceGroupInfosWithoutLock();
+    }
+}
+
+template <typename NestedQueueType>
+void ResourceControlQueue<NestedQueueType>::updateStatistics(const TaskPtr & task, ExecTaskStatus, size_t inc_value)
+{
+    assert(task);
+    const std::string & name = task->getResourceGroupName();
+
+    std::lock_guard lock(mu);
+    auto iter = resource_group_statistic.find(name);
+    if (iter == resource_group_statistic.end())
+    {
+        UInt64 accumulated_cpu_time = inc_value;
+        if (accumulated_cpu_time >= YIELD_MAX_TIME_SPENT_NS)
+        {
+            updateResourceGroupStatisticWithoutLock(name, accumulated_cpu_time);
+            accumulated_cpu_time = 0;
+        }
```

Comment:
The cache value is a bit weird...how about doing that todo directly `maybe change TaskThreadPool code, so only call updateStatistics each 100ms`

Reply:
Maybe cache the inc_value in the LAC?
In fact, the LAC itself will cache the local token. Is it meaningful to cache inc_value in the Resource Group queue?

Reply:
good idea

Reply:
1. `call updateStatistics each 100ms` this may affects mlfq
2. move `inc_value` to LAC means every time `updateStaticstics` is called, LAC need lock to add `inc_value` to `accmulated_cpu_time`. 
After discuss, leave current implementation unchanged.